Python/_warnings.c

```diff
@@ -1146,50 +1151,36 @@ static PyMethodDef warnings_functions[] = {
 static PyObject *
 create_filter(PyObject *category, const char *action)
 {
-    static PyObject *ignore_str = NULL;
-    static PyObject *error_str = NULL;
-    static PyObject *default_str = NULL;
-    static PyObject *always_str = NULL;
-    PyObject *action_obj = NULL;
+    _Py_IDENTIFIER(ignore);
+    _Py_IDENTIFIER(error);
+    _Py_IDENTIFIER(always);
+    _Py_static_string(PyId_default, "default");
+    _Py_Identifier *id;
 
     if (!strcmp(action, "ignore")) {
-        if (ignore_str == NULL) {
-            ignore_str = PyUnicode_InternFromString("ignore");
-            if (ignore_str == NULL)
-                return NULL;
-        }
-        action_obj = ignore_str;
+        id = &PyId_ignore;
     }
     else if (!strcmp(action, "error")) {
-        if (error_str == NULL) {
-            error_str = PyUnicode_InternFromString("error");
-            if (error_str == NULL)
-                return NULL;
-        }
-        action_obj = error_str;
+        id = &PyId_error;
     }
     else if (!strcmp(action, "default")) {
-        if (default_str == NULL) {
-            default_str = PyUnicode_InternFromString("default");
-            if (default_str == NULL)
-                return NULL;
-        }
-        action_obj = default_str;
+        id = &PyId_default;
     }
     else if (!strcmp(action, "always")) {
-        if (always_str == NULL) {
-            always_str = PyUnicode_InternFromString("always");
-            if (always_str == NULL)
-                return NULL;
-        }
-        action_obj = always_str;
+        id = &PyId_always;
     }
     else {
-        Py_FatalError("unknown action");
+        PyErr_SetString(PyExc_ValueError, "unknown action");
```

Comment:
Please revert this change. This is an internal function called with predefined set of arguments. The code in the `else` branch is a dead code. Here should be `Py_FatalError()` or `Py_UNREACHABLE()`.

Reply:
I like your idea of using _Py_Identifier: I proposed https://github.com/python/cpython/pull/4516 to implement your idea. It avoids the need to have to unhandle "unknown action".

FYI I'm trying to remove all Py_FatalError() calls from the Python source code, one by one, when it's possible and when it makes sense. I added a new _PyInitError API to report errors at Python initialization, so the caller can decide when Py_FatalError() will be called, and maybe release memory before calling Py_FatalError() (or flush files, or something else).